File: verifiers/federation-node/src/cbor.js
const LIMITS = { bytes: 1_048_576, depth: 24, array: 512, map: 256, string: 262_144 };
export class Tagged { constructor(tag, value) { this.tag = tag; this.value = value; } }
function fail(message) { throw new Error(`CBOR ${message}`); }
function head(major, n) {
  n = BigInt(n); const base = major << 5;
  if (n < 24n) return Buffer.from([base | Number(n)]);
  for (const [ai, width, max] of [[24,1,0xffn],[25,2,0xffffn],[26,4,0xffffffffn],[27,8,0xffffffffffffffffn]]) if (n <= max) { const b=Buffer.alloc(1+width); b[0]=base|ai; for(let i=width;i;i--){b[i]=Number(n&255n);n>>=8n;} return b; }
  fail("integer range");
}
function enc(v, opts, depth, seen) {
  if (depth > (opts.maxDepth ?? LIMITS.depth)) fail("depth limit");
  if (v === null) return Buffer.from([0xf6]); if (v === false) return Buffer.from([0xf4]); if (v === true) return Buffer.from([0xf5]);
  if (typeof v === "number" || typeof v === "bigint") { if(typeof v === "number" && !Number.isSafeInteger(v)) fail("integer type"); const n=BigInt(v); return n>=0n?head(0,n):head(1,-1n-n); }
  if (Buffer.isBuffer(v)) return Buffer.concat([head(2,v.length),v]);
  if (typeof v === "string") { const b=Buffer.from(v,"utf8"); if(b.toString("utf8")!==v) fail("UTF-8"); return Buffer.concat([head(3,b.length),b]); }
  if (seen.has(v)) fail("cycle"); seen.add(v);
  try {
    if (v instanceof Tagged) { if(!opts.allowTag18 || v.tag!==18) fail("unsupported tag"); return Buffer.concat([head(6,18),enc(v.value,opts,depth+1,seen)]); }
    if (Array.isArray(v)) { if(v.length>(opts.maxArrayItems??LIMITS.array)) fail("array limit"); return Buffer.concat([head(4,v.length),...v.map(x=>enc(x,opts,depth+1,seen))]); }
    if (v instanceof Map) { if(v.size>(opts.maxMapPairs??LIMITS.map)) fail("map limit"); const entries=[...v].map(([k,x])=>[enc(k,opts,depth+1,seen),x]).sort((a,b)=>Buffer.compare(a[0],b[0])); for(let i=1;i<entries.length;i++) if(entries[i-1][0].equals(entries[i][0])) fail("duplicate map key"); return Buffer.concat([head(5,entries.length),...entries.flatMap(([k,x])=>[k,enc(x,opts,depth+1,seen)])]); }
  } finally { seen.delete(v); }
  fail("unsupported value");
}
export function encodeCanonical(value, options={}) { const b=enc(value,options,1,new Set()); if(b.length>(options.maxBytes??LIMITS.bytes)) fail("byte limit"); return b; }
class Decoder {
  constructor(b,o){this.b=b;this.o=o;this.i=0;}
  take(n){if(this.i+n>this.b.length)fail("truncated");const x=this.b.subarray(this.i,this.i+n);this.i+=n;return x;}
  arg(ai){if(ai<24)return BigInt(ai);const widths={24:1,25:2,26:4,27:8},w=widths[ai];if(!w)fail("indefinite/reserved");let n=0n;for(const x of this.take(w))n=(n<<8n)|BigInt(x);const min={24:24n,25:256n,26:65536n,27:4294967296n}[ai];if(n<min)fail("non-shortest integer/length");return n;}
  count(n,max){if(n>BigInt(max)||n>BigInt(Number.MAX_SAFE_INTEGER))fail("resource limit");return Number(n);}
  item(d=1){if(d>(this.o.maxDepth??LIMITS.depth))fail("depth limit");const x=this.take(1)[0],m=x>>5,ai=x&31,n=()=>this.arg(ai);if(m===0){const z=n();return z<=BigInt(Number.MAX_SAFE_INTEGER)?Number(z):z;}if(m===1){const z=-1n-n();return z>=BigInt(Number.MIN_SAFE_INTEGER)?Number(z):z;}if(m===2||m===3){const z=this.count(n(),this.o.maxStringBytes??LIMITS.string),b=this.take(z);if(m===2)return Buffer.from(b);const s=new TextDecoder("utf-8",{fatal:true}).decode(b);return s;}if(m===4){const z=this.count(n(),this.o.maxArrayItems??LIMITS.array),a=[];while(a.length<z)a.push(this.item(d+1));return a;}if(m===5){const z=this.count(n(),this.o.maxMapPairs??LIMITS.map),map=new Map();let prev=null;for(let j=0;j<z;j++){const start=this.i,k=this.item(d+1),raw=this.b.subarray(start,this.i);if(prev&&Buffer.compare(prev,raw)>=0)fail("duplicate or map key order");prev=raw;map.set(k,this.item(d+1));}return map;}if(m===6){const tag=n();if(!this.o.allowTag18||tag!==18n)fail("unsupported tag");return new Tagged(18,this.item(d+1));}if(m===7&&ai===20)return false;if(m===7&&ai===21)return true;if(m===7&&ai===22)return null;fail("simple/float");}
}
export function decodeCanonical(bytes,options={}){if(!Buffer.isBuffer(bytes))fail("input type");if(bytes.length>(options.maxBytes??LIMITS.bytes))fail("byte limit");const d=new Decoder(bytes,options),v=d.item();if(d.i!==bytes.length)fail("trailing bytes");if(!encodeCanonical(v,options).equals(bytes))fail("non-canonical encoding");return v;}

File: verifiers/federation-go/internal/cbor/cbor.go
// Package cbor implements the bounded deterministic CBOR surface frozen by the profile.
package cbor

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"math"
	"sort"
	"unicode/utf8"
)

type Kind uint8

const (
	Uint Kind = iota
	Nint
	Bytes
	Text
	Array
	Map
	Bool
	Null
	Tag
)

type Pair struct{ Key, Value Value }
type Value struct {
	Kind   Kind
	Uint   uint64
	Bytes  []byte
	Text   string
	Array  []Value
	Map    []Pair
	Bool   bool
	Tag    uint64
	Tagged *Value
}
type Limits struct {
	MaxDepth          int
	MaxArray, MaxMap  int
	MaxBytes, MaxText int
}

var DefaultLimits = Limits{16, 1024, 1024, 1 << 20, 1 << 20}

func Decode(raw []byte) (Value, error) { return DecodeLimits(raw, DefaultLimits) }
func DecodeLimits(raw []byte, lim Limits) (Value, error) {
	d := decoder{b: raw, lim: lim}
	v, _, err := d.one(0)
	if err != nil {
		return Value{}, err
	}
	if d.i != len(raw) {
		return Value{}, fmt.Errorf("trailing CBOR")
	}
	return v, nil
}

type decoder struct {
	b   []byte
	i   int
	lim Limits
}

func (d *decoder) take(n int) ([]byte, error) {
	if n < 0 || n > len(d.b)-d.i {
		return nil, fmt.Errorf("truncated CBOR")
	}
	x := d.b[d.i : d.i+n]
	d.i += n
	return x, nil
}
func (d *decoder) argument(ai byte) (uint64, error) {
	if ai < 24 {
		return uint64(ai), nil
	}
	sizes := map[byte]int{24: 1, 25: 2, 26: 4, 27: 8}
	n, ok := sizes[ai]
	if !ok {
		return 0, fmt.Errorf("indefinite/reserved CBOR")
	}
	b, e := d.take(n)
	if e != nil {
		return 0, e
	}
	var x uint64
	for _, c := range b {
		x = x<<8 | uint64(c)
	}
	if x < 24 || (n > 1 && x < uint64(1)<<(8*(n-1))) {
		return 0, fmt.Errorf("non-shortest CBOR argument")
	}
	return x, nil
}
func (d *decoder) one(depth int) (Value, []byte, error) {
	if depth > d.lim.MaxDepth {
		return Value{}, nil, fmt.Errorf("CBOR depth limit")
	}
	start := d.i
	h, e := d.take(1)
	if e != nil {
		return Value{}, nil, e
	}
	major, ai := h[0]>>5, h[0]&31
	switch major {
	case 0, 1:
		x, e := d.argument(ai)
		if e != nil {
			return Value{}, nil, e
		}
		k := Uint
		if major == 1 {
			k = Nint
		}
		v := Value{Kind: k, Uint: x}
		return v, d.b[start:d.i], nil
	case 2, 3:
		n, e := d.argument(ai)
		if e != nil {
			return Value{}, nil, e
		}
		max := d.lim.MaxBytes
		if major == 3 {
			max = d.lim.MaxText
		}
		if n > uint64(max) || n > uint64(math.MaxInt) {
			return Value{}, nil, fmt.Errorf("CBOR string limit")
		}
		b, e := d.take(int(n))
		if e != nil {
			return Value{}, nil, e
		}
		if major == 2 {
			return Value{Kind: Bytes, Bytes: append([]byte(nil), b...)}, d.b[start:d.i], nil
		}
		if !utf8.Valid(b) {
			return Value{}, nil, fmt.Errorf("invalid UTF-8")
		}
		return Value{Kind: Text, Text: string(b)}, d.b[start:d.i], nil
	case 4:
		n, e := d.argument(ai)
		if e != nil {
			return Value{}, nil, e
		}
		if n > uint64(d.lim.MaxArray) {
			return Value{}, nil, fmt.Errorf("CBOR array limit")
		}
		a := make([]Value, 0, n)
		for range n {
			x, _, e := d.one(depth + 1)
			if e != nil {
				return Value{}, nil, e
			}
			a = append(a, x)
		}
		return Value{Kind: Array, Array: a}, d.b[start:d.i], nil
	case 5:
		n, e := d.argument(ai)
		if e != nil {
			return Value{}, nil, e
		}
		if n > uint64(d.lim.MaxMap) {
			return Value{}, nil, fmt.Errorf("CBOR map limit")
		}
		m := make([]Pair, 0, n)
		var prev []byte
		for range n {
			k, kb, e := d.one(depth + 1)
			if e != nil {
				return Value{}, nil, e
			}
			if prev != nil && keyCompare(prev, kb) >= 0 {
				return Value{}, nil, fmt.Errorf("duplicate or noncanonical map key")
			}
			prev = append(prev[:0], kb...)
			v, _, e := d.one(depth + 1)
			if e != nil {
				return Value{}, nil, e
			}
			m = append(m, Pair{k, v})
		}
		return Value{Kind: Map, Map: m}, d.b[start:d.i], nil
	case 6:
		t, e := d.argument(ai)
		if e != nil {
			return Value{}, nil, e
		}
		x, _, e := d.one(depth + 1)
		if e != nil {
			return Value{}, nil, e
		}
		return Value{Kind: Tag, Tag: t, Tagged: &x}, d.b[start:d.i], nil
	case 7:
		if ai == 20 {
			return Value{Kind: Bool, Bool: false}, d.b[start:d.i], nil
		}
		if ai == 21 {
			return Value{Kind: Bool, Bool: true}, d.b[start:d.i], nil
		}
		if ai == 22 {
			return Value{Kind: Null}, d.b[start:d.i], nil
		}
		return Value{}, nil, fmt.Errorf("unsupported simple/float")
	}
	return Value{}, nil, fmt.Errorf("unsupported CBOR")
}

func Encode(v Value) ([]byte, error) {
	var b bytes.Buffer
	if err := enc(&b, v); err != nil {
		return nil, err
	}
	return b.Bytes(), nil
}
func head(b *bytes.Buffer, major byte, x uint64) {
	switch {
	case x < 24:
		b.WriteByte(major<<5 | byte(x))
	case x <= 255:
		b.WriteByte(major<<5 | 24)
		b.WriteByte(byte(x))
	case x <= 65535:
		b.WriteByte(major<<5 | 25)
		var q [2]byte
		binary.BigEndian.PutUint16(q[:], uint16(x))
		b.Write(q[:])
	case x <= math.MaxUint32:
		b.WriteByte(major<<5 | 26)
		var q [4]byte
		binary.BigEndian.PutUint32(q[:], uint32(x))
		b.Write(q[:])
	default:
		b.WriteByte(major<<5 | 27)
		var q [8]byte
		binary.BigEndian.PutUint64(q[:], x)
		b.Write(q[:])
	}
}
func enc(b *bytes.Buffer, v Value) error {
	switch v.Kind {
	case Uint:
		head(b, 0, v.Uint)
	case Nint:
		head(b, 1, v.Uint)
	case Bytes:
		head(b, 2, uint64(len(v.Bytes)))
		b.Write(v.Bytes)
	case Text:
		if !utf8.ValidString(v.Text) {
			return fmt.Errorf("invalid UTF-8")
		}
		head(b, 3, uint64(len(v.Text)))
		b.WriteString(v.Text)
	case Array:
		head(b, 4, uint64(len(v.Array)))
		for _, x := range v.Array {
			if e := enc(b, x); e != nil {
				return e
			}
		}
	case Map:
		type ep struct{ k, v []byte }
		xs := make([]ep, len(v.Map))
		for i, p := range v.Map {
			var kb, vb bytes.Buffer
			if e := enc(&kb, p.Key); e != nil {
				return e
			}
			if e := enc(&vb, p.Value); e != nil {
				return e
			}
			xs[i] = ep{kb.Bytes(), vb.Bytes()}
		}
		sort.Slice(xs, func(i, j int) bool { return keyCompare(xs[i].k, xs[j].k) < 0 })
		for i := 1; i < len(xs); i++ {
			if bytes.Equal(xs[i-1].k, xs[i].k) {
				return fmt.Errorf("duplicate map key")
			}
		}
		head(b, 5, uint64(len(xs)))
		for _, x := range xs {
			b.Write(x.k)
			b.Write(x.v)
		}
	case Bool:
		if v.Bool {
			b.WriteByte(0xf5)
		} else {
			b.WriteByte(0xf4)
		}
	case Null:
		b.WriteByte(0xf6)
	case Tag:
		head(b, 6, v.Tag)
		if v.Tagged == nil {
			return fmt.Errorf("nil tag")
		}
		return enc(b, *v.Tagged)
	default:
		return fmt.Errorf("unknown CBOR kind")
	}
	return nil
}
func keyCompare(a, b []byte) int {
	if len(a) < len(b) {
		return -1
	}
	if len(a) > len(b) {
		return 1
	}
	return bytes.Compare(a, b)
}

func Int(v Value) (int64, bool) {
	if v.Kind == Uint && v.Uint <= math.MaxInt64 {
		return int64(v.Uint), true
	}
	if v.Kind == Nint && v.Uint <= math.MaxInt64 {
		return -1 - int64(v.Uint), true
	}
	return 0, false
}
func Get(m Value, key uint64) (Value, bool) {
	if m.Kind != Map {
		return Value{}, false
	}
	for _, p := range m.Map {
		if p.Key.Kind == Uint && p.Key.Uint == key {
			return p.Value, true
		}
	}
	return Value{}, false
}
